motionplan/inverseKinematics.go

```diff
@@ -2,23 +2,232 @@ package motionplan
+// by proximity to some reference input that is also specified by the user.  Finally, a WorldState argument allows users to
+// disallow or allow regions of state space through defining obstacles or interaction spaces respectively.
+func BestIKSolutions(
+	ctx context.Context,
+	logger golog.Logger,
+	fs referenceframe.FrameSystem,
+	frame referenceframe.Frame,
+	inputMap map[string][]referenceframe.Input,
+	goal spatialmath.Pose,
+	worldState *referenceframe.WorldState,
+	randseed int,
+	ikConfig map[string]interface{},
+) ([][]referenceframe.Input, error) {
+	opt, err := newPlanManager(logger, fs, frame, inputMap, goal, worldState, randseed, ikConfig).plannerOptionsFromConfig(nil, goal, ikConfig)
+	if err != nil {
+		return nil, err
+	}
+	ik, err := newIKSolver(frame, logger, opt.ikOptions)
+	if err != nil {
+		return nil, err
+	}
+	input, err := frame.InputFromMap(inputMap)
+	if err != nil {
+		return nil, err
+	}
+	nodes, err := getSolutions(ctx, ik, goal, input, randseed)
+	if err != nil {
+		return nil, err
+	}
+	solutions := make([][]referenceframe.Input, 0)
+	for _, node := range nodes {
+		solutions = append(solutions, node.Q())
+	}
+	return solutions, nil
+}
+
+func getSolutions(
+	ctx context.Context,
+	ik inverseKinematicsSolver,
+	goal spatialmath.Pose,
+	input []referenceframe.Input,
+	randseed int,
+) ([]*costNode, error) {
+	seedPos, err := ik.frame().Transform(input)
+	if err != nil {
+		return nil, err
+	}
+	goalPos := fixOvIncrement(goal, seedPos)
+
+	solutionGen := make(chan []referenceframe.Input)
+	ikErr := make(chan error, 1)
+	defer func() { <-ikErr }()
+
+	ctxWithCancel, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Spawn the IK solver to generate solutions until done
+	utils.PanicCapturingGo(func() {
+		defer close(ikErr)
+		ikErr <- ik.solve(ctxWithCancel, solutionGen, goalPos, input, ik.options().metric, randseed)
+	})
+
+	solutions := map[float64][]referenceframe.Input{}
+
+	// Solve the IK solver. Loop labels are required because `break` etc in a `select` will break only the `select`.
+IK:
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+		}
+
+		select {
+		case step := <-solutionGen:
+			cPass, cScore := ik.options().CheckConstraints(&ConstraintInput{
+				StartPos:   seedPos,
+				EndPos:     goalPos,
+				StartInput: input,
+				EndInput:   step,
+				Frame:      ik.frame(),
+			})
+			endPass, _ := ik.options().CheckConstraints(&ConstraintInput{
+				StartPos:   goalPos,
+				EndPos:     goalPos,
+				StartInput: step,
+				EndInput:   step,
+				Frame:      ik.frame(),
+			})
+
+			if cPass && endPass {
+				if cScore < ik.options().MinScore && ik.options().MinScore > 0 {
+					solutions = map[float64][]referenceframe.Input{}
+					solutions[cScore] = step
+					// good solution, stopping early
+					break IK
+				}
+
+				solutions[cScore] = step
+				if len(solutions) >= ik.options().MaxSolutions {
```

Comment:
`options` should be cached outside the for loop, rather than making a function call each iteration

Reply:
changed